lightningd/plugin.c

```diff
@@ -32,8 +49,124 @@ void plugin_register(struct plugins *plugins, const char* path TAKES)
 	p->cmd = tal_strdup(p, path);
 }
 
+/**
+ * Try to parse a complete message from the plugin's buffer.
+ *
+ * Internally calls the handler if it was able to fully parse a JSON message,
+ * and returns true in that case.
+ */
+static bool plugin_read_json_one(struct plugin *plugin)
+{
+	jsmntok_t *toks;
+	bool valid;
+	toks = json_parse_input(plugin->buffer, plugin->used, &valid);
+	if (!toks) {
+		if (!valid) {
+			/* FIXME (cdecker) Print error and kill the plugin */
+			return io_close(plugin->stdout_conn);
+		}
+		/* We need more. */
+		return false;
+	}
+
+	/* Empty buffer? (eg. just whitespace). */
+	if (tal_count(toks) == 1) {
+		plugin->used = 0;
+		return false;
+	}
+
+	/* FIXME(cdecker) Call dispatch to handle this message. */
+
+	/* Move this object out of the buffer */
+	memmove(plugin->buffer, plugin->buffer + toks[0].end,
+		tal_count(plugin->buffer) - toks[0].end);
+	plugin->used -= toks[0].end;
+	tal_free(toks);
+	return true;
+}
+
+static struct io_plan *plugin_read_json(struct io_conn *conn UNUSED,
+					struct plugin *plugin)
+{
+	bool success;
+	plugin->used += plugin->len_read;
+	if (plugin->used == tal_count(plugin->buffer))
+		tal_resize(&plugin->buffer, plugin->used * 2);
+
+	/* Read and process all messages from the connection */
+	do {
+		success = plugin_read_json_one(plugin);
+	} while (success);
+
+	/* Now read more from the connection */
+	return io_read_partial(plugin->stdout_conn,
+			       plugin->buffer + plugin->used,
+			       tal_count(plugin->buffer) - plugin->used,
+			       &plugin->len_read, plugin_read_json, plugin);
+}
+
+static struct io_plan *plugin_write_json(struct io_conn *conn UNUSED,
+					 struct plugin *plugin)
+{
+	struct json_output *out;
+	out = list_pop(&plugin->output, struct json_output, list);
+	if (!out) {
+		if (plugin->stop) {
+			return io_close(conn);
+		} else {
+			return io_out_wait(plugin->stdin_conn, plugin,
+					   plugin_write_json, plugin);
+		}
+	}
+
+	/* We have a message we'd like to send */
+	plugin->outbuf = tal_steal(plugin, out->json);
+	tal_free(out);
+	return io_write(conn, plugin->outbuf, strlen(plugin->outbuf),
+			plugin_write_json, plugin);
+}
+
+static struct io_plan *plugin_conn_init(struct io_conn *conn,
+					struct plugin *plugin)
+{
+	plugin->stop = false;
```

Comment:
I think this function should init the conn, not the plugin. Put that in the caller?